### PR TITLE
Improve hot reloading

### PR DESCRIPTION
**Pull request by TimUntersberger**
edge cases:

- [x] display_app_bar
- [x] remove_task_bar
- [x] launch_on_startup
- [x] use_border
- [x] outer_gap
- [x] inner_gap
- [x] remove_title_bar
- [x] light_theme
- [x] work_mode
- [x] multi_monitor
- [x] workspaces
- [x] bar.color
- [x] bar.height
- [x] bar.font
- [x] bar.font_size
- [x] bar.components
- [x] enable_hot_reloading

todo
- [x] state.redraw() doesnt work
- [x] handle erros in transition functions
- [x] when hot reloading keybindings don't always override the previous ones.

### CODE REVIEW

**Comment by TimUntersberger** [comment]
@ramirezmike I am currently trying to redraw all visible grids, but I can't find a way to do this.

this is how I am doing it right now.

```rust
                    "use_border" => set_prop!(use_border, bool, |old, new, _| {
                        if old != new {
                            state.each_window(|w| {
                                if new {
                                    w.add_border()
                                } else {
                                    w.remove_border()
                                }
                            }).unwrap();

                            state.config.use_border = new;
                            for d in &state.displays {
                                // this doesn't work
                                if let Some(g) = d.get_focused_grid() {
                                    d.refresh_grid(&state.config).unwrap();
                                }
                            }
                        }
                    }),

```

This is the branch of a match statement. It basically modifies the window style of all managed windows based on the new value and then it should redraw the visible grids.

**Comment by ramirezmike** [comment]
I'll try to take a closer look later tonight but you shouldn't need this part wrapping the call

```
                                if let Some(g) = d.get_focused_grid() {
                                }
```
because refresh_grid handles that. I don't think it would hurt but... it shouldn't be necessary. I would look to see if the display struct's focused_grid_id is not None.

File: twm/src/display.rs [hunk 1]
```diff
@@ -36,6 +36,23 @@ impl Display {
     pub fn get_rect(&self) -> Rectangle {
         api::get_display_rect(self.id)
     }
+    pub fn cleanup(&mut self, remove_task_bar: bool) -> SystemResult {
+        if let Some(bar) = self.appbar.as_ref() {
+            bar.window.close()?;
+        }
+
+        if remove_task_bar {
```
Comment (ramirezmike) [inline twm/src/display.rs:44]: "remove_task_bar" is almost confusing here.. like, it represents "config had remove_task_bar set", right?

would it be better as "show_task_bar" ?

File: twm/src/lua/mod.rs [hunk 2]
```diff
@@ -334,38 +334,163 @@ fn setup_nog_global(state_arc: Arc<Mutex<AppState>>, rt: &LuaRuntime) {
...
             Value,
             Value,
             Value
         )| {
-            validate!(lua, { prefix: String, key: String });
+            validate!(lua, { prefix: String, key: String, is_setup: bool });
             let parts = prefix.split('.').collect::<Vec<_>>();
             let state_arc = state.clone();
             let mut state = state_arc.lock();
             macro_rules! set_prop {
```
Comment (ramirezmike) [inline twm/src/lua/mod.rs:348]: this is pretty cool btw

File: twm/src/lua/mod.rs [hunk 3]
```diff
@@ -334,38 +334,163 @@ fn setup_nog_global(state_arc: Arc<Mutex<AppState>>, rt: &LuaRuntime) {
...
+                        if old != new && state.work_mode {
+                            state.each_window(|w| {
+                                if new {
+                                    w.add_border()
+                                } else {
+                                    w.remove_border()
+                                }
+                            }).unwrap();
+
+                            state.config.use_border = new;
```
Comment (ramirezmike) [inline twm/src/lua/mod.rs:402]: does this mean that... when you hot reload while not in work mode... then enter work mode the values aren't right? I would think that you'd always set state.config.use_border to whatever new is, right?

Comment (ramirezmike) [reply twm/src/lua/mod.rs:402]: this might apply to the rest of these too

File: twm/src/lua/mod.rs [hunk 4]
```diff
@@ -334,38 +334,163 @@ fn setup_nog_global(state_arc: Arc<Mutex<AppState>>, rt: &LuaRuntime) {
...
+                                drop(state);
+                                if display_app_bar {
+                                    AppState::close_app_bars(state_arc.clone());
+                                    AppState::create_app_bars(state_arc.clone());
+                                }
+                            }
+                        }
+                        Ok(())
+                    }),
+                    "remove_task_bar" => set_prop!(remove_task_bar, bool, |old, new, _| -> RuntimeResult<()> {
```
Comment (ramirezmike) [inline twm/src/lua/mod.rs:473]: this and display_app_bar don't update state.config?

**Review by ramirezmike — COMMENTED** [review]
I left some comments but nothing too big jumped out at me.

File: twm/src/display.rs (hunk 1 above)
Comment (TimUntersberger) [reply twm/src/display.rs:44]: true 😆

File: twm/src/lua/mod.rs (hunk 2 above)
Comment (TimUntersberger) [reply twm/src/lua/mod.rs:348]: You mean the validate! macro?

File: twm/src/lua/mod.rs (hunk 3 above)
Comment (TimUntersberger) [reply twm/src/lua/mod.rs:402]: Hmm ok this a bit confusing sorry. The set_prop! macro always sets the property of the config after the function has been called, but sometimes we need the new config value in the function itself so I have to set it manually. I will add comments for this.

File: twm/src/lua/mod.rs (hunk 4 above)
Comment (TimUntersberger) [reply twm/src/lua/mod.rs:473]: Same thing as above.

File: twm/src/lua/mod.rs (hunk 2 above)
Comment (ramirezmike) [reply twm/src/lua/mod.rs:368]: no, the set_prop macro. Although that's cool too.

**Review by ramirezmike — APPROVED** [review]
looks good to me

File: twm/src/lua/mod.rs [hunk 5]
```diff
@@ -345,6 +345,26 @@ fn setup_nog_global(state_arc: Arc<Mutex<AppState>>, rt: &LuaRuntime) {
...
+            /// This macro makes it a lot easier to update the config and handle transitions.
+            ///
+            /// The most basic usage looks like this:
+            ///
+            /// set_prop!("display_app_bar", bool)
+            ///
+            /// The above line validates the value argument as a bool and if the validation
+            /// succeeds it updates the config.
+            ///
+            /// To update the bar config you have to the same thing but with bar infront:
```
Comment (ramirezmike) [inline twm/src/lua/mod.rs:358]: it looks like you lost a word here? maybe "you have to do the same thing"?

**Comment by ramirezmike** [comment]
is this waiting on something or can it be merged?

**Comment by TimUntersberger** [comment]
I don't really remember why I didn't merge this 🤔 

Will merge this and hope I didn't forget about anything.

**Pull request merged**